### PR TITLE
add failover trigger to domain cache

### PR DESCRIPTION
**Pull request by wxing1292**
* add failover trigger to domain cache
* wire failover trigger registration && de-registration to history engine

solves #708

### CODE REVIEW

File: common/cache/domainCache.go [hunk 1]
```diff
@@ -42,12 +42,16 @@ const (
...
 type (
+	callbackFn func(domainID string, prevActiveDomain string)
+
 	// DomainCache is used the cache domain information and configuration to avoid making too many calls to cassandra.
 	// This cache is mainly used by frontend for resolving domain names to domain uuids which are used throughout the
 	// system.  Each domain entry is kept in the cache for one hour but also has an expiry of 10 seconds.  This results
 	// in updating the domain entry every 10 seconds but in the case of a cassandra failure we can still keep on serving
 	// requests using the stale entry from cache upto an hour
 	DomainCache interface {
+		SetDomainFailoverCallback(shard int, fn callbackFn)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/cache/domainCache.go]: sure

File: common/cache/domainCache.go [hunk 2]
```diff
@@ -180,6 +210,26 @@ func (c *domainCache) getDomain(key, id, name string, cache Cache) (*DomainCache
...
+func (c *domainCache) triggerDomainFailoverCallback(domainID string, prevReplicationConfig *persistence.DomainReplicationConfig,
+	nextReplicationConfig *persistence.DomainReplicationConfig) {
+
+	if prevReplicationConfig == nil || nextReplicationConfig == nil {
+		return
+	}
+	prevActiveDomain := prevReplicationConfig.ActiveClusterName
+	nextActiveDomain := nextReplicationConfig.ActiveClusterName
+
+	if prevActiveDomain == nextActiveDomain || nextActiveDomain != c.clusterMetadata.GetCurrentClusterName() {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/cache/domainCache.go]: ok

File: common/cache/domainCache.go [hunk 3]
```diff
@@ -42,12 +42,16 @@ const (
 )
 
 type (
+	callbackFn func(domainID string, prevActiveDomain string)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/cache/domainCache.go]: sure, that is a typo

File: common/cache/domainCache.go [hunk 4]
```diff
@@ -180,6 +210,26 @@ func (c *domainCache) getDomain(key, id, name string, cache Cache) (*DomainCache
...
+	nextActiveDomain := nextReplicationConfig.ActiveClusterName
+
+	if prevActiveDomain == nextActiveDomain || nextActiveDomain != c.clusterMetadata.GetCurrentClusterName() {
+		return
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+	for _, callback := range c.callbacks {
+		callback(domainID, prevActiveDomain)
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/cache/domainCache.go]: i can start a dedicated coroutine for fanout the event.

File: common/cache/domainCache.go [hunk 5]
```diff
@@ -180,6 +210,26 @@ func (c *domainCache) getDomain(key, id, name string, cache Cache) (*DomainCache
...
+		return
+	}
+	prevActiveDomain := prevReplicationConfig.ActiveClusterName
+	nextActiveDomain := nextReplicationConfig.ActiveClusterName
+
+	if prevActiveDomain == nextActiveDomain || nextActiveDomain != c.clusterMetadata.GetCurrentClusterName() {
+		return
+	}
+
+	c.RLock()
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/cache/domainCache.go]: this is kind of a write once, read all the remaining time situation.
write once: when a shard is acquired, a call back will be resigered.
read all the remaining time: when a domain is changed.

File: common/cache/domainCache.go [hunk 6]
```diff
@@ -300,3 +308,33 @@ func (entry *DomainCacheEntry) GetDomainNotActiveErr() error {
 	}
 	return errors.NewDomainNotActiveError(entry.info.Name, entry.clusterMetadata.GetCurrentClusterName(), entry.replicationConfig.ActiveClusterName)
 }
+
+// ContentEqual return true if 2 domain cache entry have the same content
+func ContentEqual(this *DomainCacheEntry, that *DomainCacheEntry) bool {
+	// we do not check the mutex or expiry
+	if *this.info != *that.info {
```
Comment (samarabbas) [inline common/cache/domainCache.go]: I thought you have a domain update version which can be used to check if the content is exactly the same?

Comment (wxing1292) [reply common/cache/domainCache.go]: that was used to deal with conflict, but that can do the work, updating

**Pull request merged**